From 11730d65d6d62711d7f8a2ce415d1229a86dd16e Mon Sep 17 00:00:00 2001
From: Peng Huang <shawn.p.huang@gmail.com>
Date: Sun, 18 Sep 2011 22:08:00 -0400
Subject: [PATCH] Fix a crash and add some warning log message.

BUG=Crash in Chrome OS
TEST=Linux desktop

Review URL: http://codereview.appspot.com/5050041
---
 src/engine.c |   19 ++++++++++++++++---
 1 files changed, 16 insertions(+), 3 deletions(-)

diff --git a/src/engine.c b/src/engine.c
index dcff0c7..db14607 100644
--- a/src/engine.c
+++ b/src/engine.c
@@ -816,8 +816,15 @@ ibus_m17n_engine_update_lookup_table (IBusM17NEngine *m17n)
             ibus_lookup_table_set_page_size (m17n->table, mtext_len (mt));
 
             buf = ibus_m17n_mtext_to_ucs4 (mt, &nchars);
-            for (i = 0; i < nchars; i++) {
-                ibus_lookup_table_append_candidate (m17n->table, ibus_text_new_from_unichar (buf[i]));
+            g_warn_if_fail (buf != NULL);
+
+            for (i = 0; buf != NULL && i < nchars; i++) {
+                IBusText *text = ibus_text_new_from_unichar (buf[i]);
+                if (text == NULL) {
+                    text = ibus_text_new_from_printf ("INVCODE=U+%04"G_GINT32_FORMAT"X", buf[i]);
+                    g_warn_if_reached ();
+                }
+                ibus_lookup_table_append_candidate (m17n->table, text);
             }
             g_free (buf);
         }
@@ -834,9 +841,15 @@ ibus_m17n_engine_update_lookup_table (IBusM17NEngine *m17n)
                 mtext = (MText *) mplist_value (p);
                 buf = ibus_m17n_mtext_to_utf8 (mtext);
                 if (buf) {
-                    ibus_lookup_table_append_candidate (m17n->table, ibus_text_new_from_string (buf));
+                    ibus_lookup_table_append_candidate (m17n->table,
+                        ibus_text_new_from_string (buf));
                     g_free (buf);
                 }
+                else {
+                    ibus_lookup_table_append_candidate (m17n->table,
+                        ibus_text_new_from_static_string ("NULL"));
+                    g_warn_if_reached();
+                }
             }
         }
 
-- 
1.7.1

